File: FindCycle.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;
        boolean cycle = false;
        while(fast != null && fast.next != null){
            slow = slow.next;
            fast = fast.next.next;
            
            if(slow == fast){
                cycle = true;
                break;
            }
        }
        if(!cycle) return null;
        fast = head;
        while(slow != fast && cycle){
            slow = slow.next;
            fast = fast.next;
        }
        return slow;
    }
}

File: findcycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        
        if(head == NULL) return NULL;
        
        ListNode* slow = head, *fast = head;
        bool flag = false; 
        while(true){
            
            
            if(slow->next == NULL || fast->next == NULL || fast->next->next == NULL) {
                flag = false;
                break;
            }
           
            slow = slow->next;
            fast = fast->next->next;
             if(slow == fast) {
                flag = true;
                break;
            }
            
            
        }
        if(flag == false) return NULL;
        cout<<fast->val<< " "<< slow->val;
        while(true){
            
            if(head == fast ) return head;
            
            head = head->next;
            fast = fast ->next;            
        }
        
        return NULL;
            
        
    }
};
